cl/api/utils.py

```diff
@@ -346,6 +349,154 @@ def allow_request(self, request, view):
+
+        self.key = self.get_cache_key_for_citations(request, view)
+        self.history = self.cache.get(self.key, [])
+        self.request_timestamp = self.timer()
+
+        # Drop any requests from the history which have now passed the
+        # throttle duration
+        while self.history and self.history[-1][-1] <= self.request_timestamp:
+            self.history.pop()
+
+        citations_in_history = sum(
+            citation_count for citation_count, timestamps in self.history
+        )
+
+        if citations_in_history >= max_num_citations:
+            self.throttle_request(request)
+
+        self.save_citation_count(request, view)
+
+    def save_citation_count(self, request, view):
+        """
+        Inserts the number of citations and the expiration time along with
+        the key into the cache.
+        """
+        citation_count = self.get_citation_count_from_request(request, view)
+        if not citation_count:
+            return
+
+        max_num_citations, duration = self.parse_rate(
+            self.get_citations_rate(request)
+        )
+        expiration = (
+            citation_count * (duration / max_num_citations)
+            if citation_count > max_num_citations
+            else duration
+        )
+        self.history.insert(
+            0,
+            [
+                citation_count,
+                self.request_timestamp + expiration,
+            ],
+        )
+
+        self.cache.set(self.key, self.history, expiration)
+
+    def allow_request(self, request, view):
+        self.throttle_request_by_citation_count(request, view)
+        return super().allow_request(request, view)
+
+    def throttle_request(self, request):
+        """
+        This helper iterates through the request history in reverse
+        chronological order to calculate the soonest time a new request can be
+        made and raises the `Throttled` exception with the details.
+
+        The exception includes details about the throttling:
+
+        - `error_message`: A message indicating the request was throttled.
+        - `wait_until`: An ISO 8601 formatted string representing the soonest
+                    time the next request can be made without throttling.
+
+        Args:
+            request: The request object to be throttled.
+
+        Raises:
+            Throttled: The exception includes details about the throttling.
+        """
+        rate = self.get_citations_rate(request)
+        max_num_citations, _ = self.parse_rate(rate)
+        soonest_time = None
+        for idx in reversed(range(len(self.history))):
+            remaining_citation = sum(
+                citation_count for citation_count, _ in self.history[:idx]
+            )
+            if remaining_citation < max_num_citations or not idx:
+                datetime_obj = datetime.fromtimestamp(
+                    self.history[idx][-1], timezone.utc
+                )
+                soonest_time = datetime_obj.isoformat()
+
```

Comment:
Shouldn't the `soonest_time` indicate the closest time a user can make a new request after history records that exceeded their quota have expired?  However, during my testing, it appears the `wait_until` is not aligned with this behavior. Instead of showing the earliest time a user can submit a new request, it displays the latest time after all the user's history records have expired.

If my assumption is correct this can be easy fixed by breaking the for loop once `soonest_time` is found.

```suggestion
break
```

Reply:
You're right!: Thanks for catching that.